subsys/storage/flash_map/Kconfig

```
@@ -58,7 +58,6 @@ config FLASH_AREA_CHECK_INTEGRITY_PSA
 config FLASH_AREA_CHECK_INTEGRITY_MBEDTLS
 	bool "Use Mbed TLS"
 	select MBEDTLS
-	select MBEDTLS_SHA256
```

based on the description you need to use an if here so it is still selected when not using your config file

---

Agreeing here, you could use [the condition that guards `MBEDTLS_SHA256`](https://github.com/zephyrproject-rtos/zephyr/blob/e8589d232b0d050edbb8709aab8391a50eb0b3b1/modules/mbedtls/Kconfig.tls-generic#L8).
Though it's not great if/when e.g. we rename that file... Maybe a helper Kconfig option could be added there, which would be used here.
So this would give (in `Kconfig.tls-generic`) (not sure about the name):
```
config MBEDTLS_ZEPHYR_KCONFIGS
	def_bool MBEDTLS_BUILTIN && MBEDTLS_CFG_FILE = "config-tls-generic.h"

menu "Mbed TLS configuration"
	depends on MBEDTLS_ZEPHYR_KCONFIGS
```
And here:
```suggestion
	select MBEDTLS_SHA256 if MBEDTLS_ZEPHYR_KCONFIGS
```
@valeriosetti thoughts?

---

imply would be wrong because a user in the default configuration can disable the option and completely break the build, they should not be able to deselect a Kconfig if that Kconfig is required by another option

---

> What if we just replace select with imply here? This way just adding CONFIG_MBEDTLS_SHA256=n to the *.conf file should do the trick.

In addition to @nordicjm's point, it also feels like a confusing UX to need to have `CONFIG_MBEDTLS_SHA256=n` in the conf when you do in fact need sha256 support from mbedtls.

I believe what we really want here as an end result is that `CONFIG_MBEDTLS_SHA256_C` is enabled whether using the Zephyr default config or some other config. However, "punching through" would likely be even more confusing (i.e. I beleive setting `CONFIG_MBEDTLS_SHA256=n` when using default config would result in sha256 support still being enabled).

> My fear is that if we add this if MBEDTLS_ZEPHYR_KCONFIGS here, then we should do the same for all the occurrences of legacy CONFIG_MBEDTLS_xxx Kconfigs in the Zephyr codebase, but this is both a lot of work and not very userfriendly to read IMO.

This does sound onerous, but conceptually it feels correct (i.e. only use the Zephyr mbedtls Kconfig controls if you are using the default Zephyr mbedtls config). If the primary blocker is just the amount of work, I am happy to contribute here. I have talked to more than a handful of users who are initially confused by the Zephyr Kconfigs. I expect the churn in general is somewhat undesirable though.

---

> This does sound onerous, but conceptually it feels correct (i.e. only use the Zephyr mbedtls Kconfig controls if you are using the default Zephyr mbedtls config).

Yes, it is correct, but it would look weird to me to have `if MBEDTLS_ZEPHYR_KCONFIGS` spread all around the Zephyr codebase for all the Mbed TLS Kconfigs.

Here's my other 2 cents for the discussion.
Mbed TLS is built using configuration header files which defines desired TLS, X.509 and legacy crypto features. Zephyr instead uses Kconfig to define what's to be included in the build. `Kconfig.tls-generic` + `config-tls-generic.h` is where "the magic happens" and Zephyr's Kconfigs are tranformed in build symbols to be used by Mbed TLS. In an optimal solution, you would have a 1:1 match between Kconfigs and build symbols, so that you can specify what you prefer [^1] at Kconfig level without any need to use a custom configuration header file.
First benefits of this approach that I can think of are:
1. in case something is not working (either build time or runtime) you can easily check what's included in the build by looking at the genenerated `.config`file. Instead if you are using a custom header file then you'll have to check both Kconfigs and the header file. 
2. is more flexible in the sense that if you add some new feature in Zephyr that for some reason adds new crypto Kconfigs requirements, if we follow your approach then you'll have to add all the new build symbols to the configuration header file manually, instead with my approach everything will just be fine.
3. we don't have `if MBEDTLS_ZEPHYR_KCONFIGS` everywhere.

For some reason though, Zephyr has `CONFIG_MBEDTLS_CFG_FILE` which allows you to override `config-tls-generic.h` and to me this is not correct. What we should aim for is to have the 1:1 matching I was mentioning above.
My suspect you are not using `config-tls-generic.h` because it adds something that you cannot disable through Kconfig. Is that the case? If so, why not improving Kconfigs, adding what's missing, instead of adding another quick fix?

Besides I just checked Zephyr's main repo and I didn't find samples/tests using `CONFIG_MBEDTLS_CFG_FILE`. It's only used in a few places to get the proper configuration header file that is being used. So potentially this Kconfig can also be removed.
Moreover I've been a reviewer for Mbed TLS for some month now and this is the first time that I see this kind of complain so I assume most of the people is OK with the `Kconfig.tls-generic` + `config-tls-generic.h` combo. That's why I suggest to improve it instead of finding alternative solutions.

Since this discussion is touching Mbed TLS integration in general, let me add maintainers @ceolin @d3zd3z to the discussion so that we can have their point of view on this.

[^1]: well, there are constraints between Mbed TLS  build symbols, but that's another story.

---

Thanks for your response! I think the core issue here is that Zephyr's default mbedtls config is a leaky abstraction. Ideally flash map integrity check should be saying "I need sha256 support" and whatever crypto implementation is enabled should be configured appropriately (this requires a well defined interface that both the crypto implementation and the component are targeting -- as can be seen here when using PSA). However, when using mbedtls, there is not an interface that both are targeting; mbedtls has defined configuration and Zephyr is consuming it. When this is the case, I typically find it a better solution to be "honest" about the 1-way relationship. In this case one way of doing so would be the 1:1 Kconfig <-> mbedtls config you mention, which would wholly integrate the mbedtls config machinery into the Zephyr config machinery. The other end would be to defer to the mbedtls config system directly (which comes from the separate config systems drawback you mention).

Right now, Zephyr's semi-opinionated default mbedtls config puts users at somewhat of a middle ground. It has the positive attribute of making it such that most of the ecosystem builds "out of the box". However, the existence of the ability to override the default configuration feels like an acknowledgement that a semi-opinionated approach is not going to work for everyone, and an escape hatch might be necessary. Unfortunately, for that escape hatch to actually work, it needs to be bubbled up to components that depend on mbedtls. With the flash map integrity check, we can see that it doesn't. Using `select` ensures that a user is not able to break the build when using the default config (as @nordicjm points out), but it _prevents_ a successful build if a user is not.

Perhaps in hindsight creating a 1:1 mapping of the Zephyr mbedtls default config to the mbedtls config would be the right call, but I don't think that is feasible at this point (would potentially break some existing config), and I would honestly not recommend trying to do so unless the generation of the mapping was automated. Sure, a user could make an update to the Zephyr mbedtls default config to support their use case, but that could eventually cause conflicts, and it is an unfortunate state if the user then needs to wait until the next Zephyr release to consume it.

While it is somewhat messy, I think @tomi-font's suggestion (`MBEDTLS_ZEPHYR_KCONFIGS`) is the closest to reflecting the actual state of configuration. Furthermore, though it would probably be best to support across the board, it _could_ be added incrementally (i.e. existing config would work as it currently does without the `if MBEDTLS_ZEPHYR_KCONFIGS`).

---

Thanks for your reply :) 
IMO we both have valid points on our ends, but we sit on different sides: I'm more for improving the Kconfig mapping of Mbed TLS in Zephyr, you are more for carrying on with what we currently have and patch where necessary. 
After all also your approach might be OK. The next Mbed TLS release is expected in [September this year](https://lists.trustedfirmware.org/archives/list/mbed-tls@lists.trustedfirmware.org/thread/3MM3XIPGVAIUSX7PJ7BDILSNFRPXHBC5/) and in that one the support for legacy crypto will be removed in favor of PSA, so fix of this kind won't be needed for long time I hope.
It's basically up to the maintainer decide what to do now.

---

That's something I was thinking of suggesting: we could introduce the helper Kconfig option I mentioned, and just add it as a condition to `FLASH_AREA_CHECK_INTEGRITY_MBEDTLS`.
Then if/when further fixes are needed those can be added in due time.
It's a sort of middle ground between doing the right thing [^1] and getting something in without having to rework the whole codebase.

[^1]: Although I agree with @valeriosetti's approach where ideally we wouldn't even have the option of using a different Mbed TLS configuration file.